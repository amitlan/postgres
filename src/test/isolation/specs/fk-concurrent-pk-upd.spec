# Tests that an INSERT on referencing table correctly fails when
# the referenced value disappears due to a concurrent update
setup
{
  CREATE TABLE parent (
	parent_key	int		PRIMARY KEY,
	aux			text	NOT NULL
  );

  CREATE TABLE child (
	child_key	int		PRIMARY KEY,
	parent_key	int		NOT NULL REFERENCES parent
  );

  INSERT INTO parent VALUES (1, 'foo');
}

teardown
{
  DROP TABLE parent, child;
}

session s1
setup		{ BEGIN; }
step s1i	{ INSERT INTO child VALUES (1, 1); }
step s1c	{ COMMIT; }
step s1s	{ SELECT * FROM child; }

session s2
setup		{ BEGIN; }
step s2ukey	{ UPDATE parent SET parent_key = 2 WHERE parent_key = 1; }
step s2uaux	{ UPDATE parent SET aux = 'bar' WHERE parent_key = 1; }
step s2ukey2	{ UPDATE parent SET parent_key = 1 WHERE parent_key = 2; }
step s2c	{ COMMIT; }
step s2s	{ SELECT * FROM parent; }

# fail
permutation s2ukey s1i s2c s1c s2s s1s
# ok
permutation s2uaux s1i s2c s1c s2s s1s
# ok
permutation s2ukey s1i s2ukey2 s2c s1c s2s s1s
