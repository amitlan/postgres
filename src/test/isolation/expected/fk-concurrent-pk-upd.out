Parsed test spec with 2 sessions

starting permutation: s2ukey s1i s2c s1c s2s s1s
step s2ukey: UPDATE parent SET parent_key = 2 WHERE parent_key = 1;
step s1i: INSERT INTO child VALUES (1, 1); <waiting ...>
step s2c: COMMIT;
step s1i: <... completed>
ERROR:  insert or update on table "child" violates foreign key constraint "child_parent_key_fkey"
step s1c: COMMIT;
step s2s: SELECT * FROM parent;
parent_key|aux
----------+---
         2|foo
(1 row)

step s1s: SELECT * FROM child;
child_key|parent_key
---------+----------
(0 rows)


starting permutation: s2uaux s1i s2c s1c s2s s1s
step s2uaux: UPDATE parent SET aux = 'bar' WHERE parent_key = 1;
step s1i: INSERT INTO child VALUES (1, 1);
step s2c: COMMIT;
step s1c: COMMIT;
step s2s: SELECT * FROM parent;
parent_key|aux
----------+---
         1|bar
(1 row)

step s1s: SELECT * FROM child;
child_key|parent_key
---------+----------
        1|         1
(1 row)


starting permutation: s2ukey s1i s2ukey2 s2c s1c s2s s1s
step s2ukey: UPDATE parent SET parent_key = 2 WHERE parent_key = 1;
step s1i: INSERT INTO child VALUES (1, 1); <waiting ...>
step s2ukey2: UPDATE parent SET parent_key = 1 WHERE parent_key = 2;
step s2c: COMMIT;
step s1i: <... completed>
step s1c: COMMIT;
step s2s: SELECT * FROM parent;
parent_key|aux
----------+---
         1|foo
(1 row)

step s1s: SELECT * FROM child;
child_key|parent_key
---------+----------
        1|         1
(1 row)

